.github/workflows/pinned_build.yaml

```diff
@@ -0,0 +1,63 @@
+name: "Pinned Build"
+
+on:
+  workflow_dispatch:
+
+permissions:
+  packages: read
+  contents: read
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  Build:
+    name: Build
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: [ubuntu18, ubuntu20, ubuntu22]
+    # runs-on: ["self-hosted", "enf-x86-beefy"] # not sure if this warrants a different machine, but start here due to building
+    runs-on: ubuntu-latest
+    container: ${{ matrix.platform == 'ubuntu18' && 'ubuntu:bionic' || matrix.platform == 'ubuntu20' && 'ubuntu:focal' || 'ubuntu:jammy' }}
+    steps:
+      - name: Conditionally update git repo
+        if: ${{ matrix.platform == 'ubuntu18' }}
+        run: |
+          apt-get update
+          apt-get install -y software-properties-common
+          apt-get update
+          add-apt-repository ppa:git-core/ppa
+      - name: Update and Install git
+        run: |
+          apt-get update
+          apt-get install -y git
+          git --version
+      - name: Clone leap
+        uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Install dependencies
+        run: |
+          # https://github.com/actions/runner/issues/2033
+          chown -R $(id -u):$(id -g) $PWD
+          ./scripts/install_deps.sh
+      - name: Build Pinned Build
+        run: |
+          ./scripts/pinned_build.sh deps build "$(nproc)"
+      - name: Check and select artifact
+        run: |
+          cd build
+          echo "PINNED_BUILD_ARTIFACT="$(ls | grep -E 'leap-[^d]+*.deb')"" >> "$GITHUB_ENV"
+          echo "PINNED_BUILD_NAME="$(ls | grep -E 'leap-[^d]+*.deb' | sed -e 's/-x86_64.deb/-pinned-amd64/')"" >> "$GITHUB_ENV"
```

Comment:
What's the need for these shenanigans? I know the 3.1's .deb files are named a little differently which makes the difference between `leap` & `leap-dev` trickier. What stops us from something like this for 3.1,
```yaml
      - name: Upload package
        uses: actions/upload-artifact@v3
        with:
          name: leap-${{matrix.platform}}-pinned-amd64
          path: build/leap-3*.deb
```
and then we can make it more generic on 3.2+ with
```yaml
      - name: Upload package
        uses: actions/upload-artifact@v3
        with:
          name: leap-${{matrix.platform}}-pinned-amd64
          path: build/leap_*.deb
```

Reply:
Those shenanigans allow for selecting the `.deb` file a little more generically without having to specify the leading major version number (The wildcard matching for `upload-artifact` didn't seem to support full regex) and then crafting upload names that matched more closely what I saw in the artifacts on our releases which would include the version numbers, but then add in the `pinned` identifier:

Release assets are named as such:
![image](https://github.com/AntelopeIO/leap/assets/11048242/b33868cc-442c-4927-aadc-613f86a129c6)

The pinned build upload assets are named:
![image](https://github.com/AntelopeIO/leap/assets/11048242/cce52f78-02b2-4fb0-a084-115e007b968b)

Reply:
I don't think it's necessary for the artifact name to have the version number in it, and quite possibly we _shouldn't_ do that. It makes things harder when we want to make use of a build that isn't a release; a perfect example of that is in the libtester workflow where you simply grab the `cdt.deb` from main: what version number would you expect to find in that situation? It's difficult to know ahead of time, and then coupled with `asset-artifact-download-action` not supporting wildcards for the `artifact-name` name currently, it seems to break the use case of another repo's workflow being able to find and use any pinned build.

Admittedly once we do #1101 we're going to have to iterate over the artifacts in a workflow and look for some pattern.. but that pattern likely wouldn't need the version numbers still.

> selecting the .deb file a little more generically without having to specify the leading major version number (The wildcard matching for upload-artifact didn't seem to support full regex

afaik this is strictly a 3.1 problem, and shouldn't be a problem 3.2+, so I don't see any real problem with hardcoding the "3" in the 3.1 branch's workflow. Would the yaml samples in my above comment work?

Reply:
Sure. That's fine. I can do it that way and follow that naming convention if you'd prefer.  Would you like the artifact to have the `.deb` file extension in the name?

Reply:
![image](https://github.com/AntelopeIO/leap/assets/11048242/e2d72470-6523-4b51-88af-1925f54bcae2)